### DISCUSSION TITLE
grouping spans

### DISCUSSION THREAD

**Discussion by delanym** [open-telemetry/opentelemetry-java-contrib#317] [Q&A]
How do I get the spans produced by my javaagent running in the maven-surefire-plugin to be associated with those produced by the maven-extension?

I thought I could just set service.name to be the same, ie. my javaagent with `-Dotel.resource.attributes=service.name=maven`
matches the extension
`-Dotel.service.name=maven`
but the spans are still not collated.

**Comment by trask** [comment]
hi @delanym! IIRC the maven extension configures and explicitly uses the OTel SDK, and the Javaagent only intercepts (and collates) usage of the OTel API through GlobalOpenTelemetry object.

cc @cyrille-leclerc

**Comment by cyrille-leclerc** [comment]
Hello @delanym .
What do you want to achieve with spans on the unit tests? 
Is it a goal to measure the duration and outcome of unit test executions? To identify flaky tests and slow tests?
Or do you have the goal to fully instrument the internals of the code unit tests, including spans on the jdbc calls...?

If it's the former, tracking unit tests duration and success, then 
* it's in our radar but not solved yet. Unfortunately, the maven-surefire-plugin doesn't expose the APIs to let us do this for the moment.
* @mdelapenya has worked on this topic but I'm not sure what's the status

**Comment by delanym** [comment]
@cyrille-leclerc my involvement in the project Im working on has been from the build system inward. I've exhausted the benefits of tracing the build itself, and now Im moving steadily inward, starting with the unit tests, but with a view to fully instrumenting the production code where necessary.

For the moment I don't know the code Im working on well enough (or even the language itself) to see where the potential improvements could be. So my approach is to try instrument literally everything - the "shotgun" approach. I'm hoping the tracings can help me learn more about how it all operates.

As it happens, there are both production performance issues, and unit test JVM timeouts (I've had to set surefire property forkedProcessTimeoutInSeconds=1200 - far too long).

Im glad to hear its on your radar! Is there a surefire ticket I can follow?

@trask what is the relationship between Glowroot and OT?

**Reply by trask** [reply]
> @trask what is the relationship between Glowroot and OT?

none today. someday I hope to power the same Glowroot experiences with the OpenTelemetry bytecode instrumentation, since that supports many more libraries and has a much more active community effort

**Reply by mdelapenya** [reply]
@delanym we follow the path of instrumenting the production code (of a test framework) and realised that it comes with a lot of glue code here and there.

Then we moved to the idea of instrumenting the test runners directly, i.e. the [pytest](https://github.com/elastic/apm-pipeline-library/tree/main/resources/scripts/pytest_otel) test runner, but although possible, there are too many test runners in so many languages to contribute the instrumentation. Finally we took a shortcut to "instrument" the XML test report produced by JUnit, because we consider the XML file as a de-facto standard, having the major test runners a flag/command to export the test results to xUnit, we created two things (still in alpha!):
- a Go binary parsing the XML file and sending traces and metrics for each entity in the test report https://github.com/mdelapenya/junit2otlp
- a Jenkins plugin that does the same extending Jenkins' junit plugin (this plugin is still private)

We are consuming the Go binary [in our internal pipelines](https://github.com/elastic/apm-pipeline-library/blob/main/vars/junit2otel.groovy), and it's in our plans to create a github action doing exactly the same.

We have talked about our approach at [testguild's Automation conference](https://guildconferences.com/automation-guild-2022), and hopefully at [cdCon 2022](https://cdcon2022.sched.com/event/11Zds)
